dango/account-margin/src/query.rs

```diff
@@ -1,16 +1,86 @@
 use {
+    anyhow::bail,
     dango_auth::NEXT_SEQUENCE,
-    dango_types::account::single::QueryMsg,
-    grug::{ImmutableCtx, Json, JsonSerExt, StdResult, Storage},
+    dango_lending::DEBTS,
+    dango_oracle::OracleQueries,
+    dango_types::{
+        account::margin::{CollateralPower, HealthResponse, QueryMsg},
+        config::AppConfig,
+    },
+    grug::{
+        Addr, BorshDeExt, Coins, Denom, ImmutableCtx, Inner, IsZero, Json, JsonSerExt, NumberConst,
+        QuerierWrapper, StdResult, Storage, Udec128,
+    },
+    std::collections::BTreeMap,
 };
 
 #[cfg_attr(not(feature = "library"), grug::export)]
-pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> StdResult<Json> {
+pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> anyhow::Result<Json> {
     match msg {
         QueryMsg::Sequence {} => query_sequence(ctx.storage)?.to_json_value(),
+        QueryMsg::Health {} => query_health(ctx)?.to_json_value(),
     }
+    .map_err(Into::into)
 }
 
 fn query_sequence(storage: &dyn Storage) -> StdResult<u32> {
     NEXT_SEQUENCE.current(storage)
 }
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
+    let app_cfg: AppConfig = querier.query_app_config()?;
+
+    // Calculate the total value of the debts.
+    let total_debt_value = debts.into_iter().try_fold(Udec128::ZERO, |acc, coin| {
+        let price = querier.query_price(app_cfg.addresses.oracle, &coin.denom)?;
+        Ok::<_, anyhow::Error>(acc + price.value_of_unit_amount(coin.amount)?)
+    })?;
+
+    // Calculate the total value of the account's collateral adjusted for the collateral power.
+    let mut total_adjusted_collateral_value = Udec128::ZERO;
+    for (denom, power) in collateral_powers {
+        let collateral_balance = querier.query_balance(margin_account, denom.clone())?;
+
+        // As an optimization, don't query the price if the collateral balance is zero.
+        if collateral_balance.is_zero() {
+            continue;
+        }
+
+        let price = querier.query_price(app_cfg.addresses.oracle, &denom)?;
+        let collateral_value = price.value_of_unit_amount(collateral_balance)?;
+        total_adjusted_collateral_value += collateral_value * power.into_inner();
+    }
+
+    if total_adjusted_collateral_value.is_zero() {
+        bail!("the account has no collateral");
+    }
```

Comment:
If an account has no collateral but also no debt, it's considered healthy. In this case should return a utilization rate of zero instead of throwing error here.

Reply:
This:

```rust
let utilization_rate = if total_debt_value.is_zero() {
    Udec128::ZERO
} else if total_adjusted_collateral_value.is_zero() {
    Udec128::MAX
} else {
    total_debt_value / total_adjusted_collateral_value
};
```